### PR TITLE
Feat(app): add capacitor support, fixes #4388

### LINKED ISSUE #4388: [Big Feature] Replace Cordova with Capacitor
With an eye of developers, Capacitor has many advantages over Cordova...
Please look here:
https://capacitor.ionicframework.com/docs/cordova

It also supports Cordova plugins so it's easy to be migrated

### PR DESCRIPTION
**Pull request by nklayman**
**What kind of change does this PR introduce?** (check at least one)

- [x] Feature

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch and _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] It's been tested on a Cordova (iOS, Android) app
- [ ] It's been tested on a Electron app
- [x] Any necessary documentation has been added or updated [in the docs](https://github.com/quasarframework/quasar/tree/dev/docs) (for faster update click on "Suggest an edit on GitHub" at bottom of page) or explained in the PR's description.

**Other information:**
Have not tested on mac or windows. To test:
1. Clone fork, cd into `app`, install deps, link with `yarn link`
2. create new project with `quasar create capacitor-quasar` (use yarn), or use existing project (backup first!)
3. link in project with `yarn link @quasar/app`
4. add mode with `quasar mode add capacitor`, and choose your platforms
5. dev with `quasar dev -m capacitor --target [android|ios]`
6. build with `quasar build -m capacitor --target [android|ios]`

### CODE REVIEW

**Comment by rstoenescu** [comment]
Hi Noah,

Excellent work! Few things:

1. In order to be consistent with the folder structure, can you place the new folders in a `/src-capacitor` instead?
2. Is there a way for us not to hard-code paths to capacitor executable? I'm worried about working with monorepos as this would break it.

**Comment by nklayman** [comment]
1. I don't think Capacitor has support for a different folder location, other than the root. It would be pretty complicated to get around it. You could open an issue in the capacitor repository to try and solve this.
2. [execa](https://github.com/sindresorhus/execa) has support for [executing locally installed binaries](https://github.com/sindresorhus/execa#preferlocal). I'm not sure that it would support yarn workspaces either, but it might.

**Comment by lucasfernog** [comment]
Cant' you use "src-capacitor" as root, calling capacitor commands using appPaths.capacitor as cwd @nklayman ?

**Comment by lucasfernog** [comment]
About the execution of locally installed capacitor, can't you use something like what Electron is doing: https://github.com/quasarframework/quasar/blob/dev/app/lib/electron/index.js#L161 ?

**Comment by nklayman** [comment]
@lucasfernog using `src-capacitor` as the root could work, but capacitor would try to install npm deps there. They could possibly be prevented, but I'm not sure how to do that. As for executing the binaries, that's a great idea. I'll go ahead and implement it now. Using execa is still a good option, though, as it has a bunch of extra improvements over `child_process`.

**Comment by nklayman** [comment]
I added the `require.resolve` binary locating method, as well as a fallback to the old method as `require.resolve` doesn't work when using `yarn link @quasar/app`.

**Review by nothingismagick — CHANGES_REQUESTED** [review]
On MacOS 
`yarn cap add ios` => 
[error] cocoapods is not installed. For information: https://guides.cocoapods.org/using/getting-started.html#installation

Cocoapods requires ruby. Installed with:
`brew install ruby`
Ruby brings these deps:
==> Installing dependencies for ruby: libyaml, openssl and readline

CocoaPods installed with:
`sudo gem install cocoapods`
CocoaPods brings these deps:
thread_safe, tzinfo, concurrent-ruby, i18n, activesupport, nap, fuzzy_match, cocoapods-core, claide, cocoapods-deintegrate, cocoapods-downloader, cocoapods-plugins, cocoapods-search, cocoapods-stats, netrc, cocoapods-trunk, cocoapods-try, molinillo, atomos, CFPropertyList, colored2, nanaimo, xcodeproj, escape, fourflusher, gh_inspector, ruby-macho

After installing all deps and rerunning the cmd, this is the error:
[error] Capacitor could not find the web assets directory "/[redacted]/capacitor-quasar/dist/capacitor".

So I then run:
`quasar build -m capacitor --target ios` 

Which does create the assets, but still throws an error:
[error] ios" platform has not been created. Use "npx cap add ios" to add the platform project.

So then I run `yarn cap add ios` and this works now. 
It spends a bit of time (in my case still running the update loop after more than 15 minutes...) to update iOS native dependencies. Unfortunately this is all I see:
```
Updating iOS native dependencies with "pod install" (may take several minutes)
```
So there is ZERO information about what is being installed on my system, whether the cmd has hung etc. This is a little worrisome from a security standpoint, because I had to install cocoapods with sudo. Maybe it will tell me what it did at the end of its loop (if it ends)...

Will report back when it finally finishes. Note - I have not yet had the opportunity to run `quasar dev -m capacitor --target ios`

I stopped waiting and cancelled. Now, regardless of whether I try `dev` or `build` it hangs at Updating iOS native dependencies.

So I cd into `ios/App` and run `$ pod install`
It is downloading a really MASSIVE spec (ca 750MB) from here:
https://github.com/CocoaPods/Specs.git

```
CocoaPods is a dependency manager for Swift and Objective-C Cocoa projects. It has over 63 thousand libraries and is used in over 3 million apps. CocoaPods can help you scale your projects elegantly.
```
See this about saving time and storage:
http://blog.cocoapods.org/CocoaPods-1.7.2/

---

It did finally spawn xcode, but the xcode build failed on both dev and build. Can't triage at the moment.

**Comment by lucasfernog** [comment]
@nklayman About the src-capacitor issue, would this work? https://github.com/ionic-team/capacitor/issues/691#issuecomment-452384208

Sorry i don't know much about it beucase i never used Capacitor, i'm just blindly searching for a solution.

**Comment by lucasfernog** [comment]
I'm testing this and I found that `capacitor init` prompts me about using NPM or Yarn. It would be nice if you detected it before calling the command and uses the flag "--npm-client" from the command.

**Comment by lucasfernog** [comment]
The default Android Studio path from Capacitor is not the one I have on my Arch Linux machine.. should we try to read it for users, and change the capacitor.config.json with it?

File: app/lib/capacitor/index.js
Comment (lucasfernog) [inline app/lib/capacitor/index.js]: I believe this is not always true.. Quasar picked up my local network IP as url... I dont know

**Comment by lucasfernog** [comment]
I think we can use our own templates for the android/ and ios/ folders... it would be SUPER COOL, it would be nice to use lodash templates and get rid of the cleartext validation

**Comment by lucasfernog** [comment]
Screenshot_1561670998

Android worked easily on an emulator (already had Android installed)... Amazing job @nklayman !!

**Comment by nklayman** [comment]
@nothingismagick the cocoapods issue is with capacitor (https://github.com/ionic-team/capacitor/issues/626#issuecomment-401533991). You could open an issue on their repo to provide more helpful output when running the command. Not much I can do since I don't have a Mac.

@lucasfernog I made pr on capacitor to fix the npm-client prompt issue, since I added support for yarn to capacitor in the first place. It was already merged, and should be part of the next release. What do you mean by "using your own template for `android` and `ios`"? Android studio path detection on Linux has been discussed on capacitor at https://github.com/ionic-team/capacitor/issues/560. I could try and fix it and open a PR to capacitor. Also, go Arch/Linux :+1:!

File: app/lib/capacitor/index.js
Comment (nklayman) [reply app/lib/capacitor/index.js]: @rstoenescu what is the best way to test if the Quasar dev server url is on the network or just a localhost url?

**Comment by lucasfernog** [comment]
The contents that gets copied to /android is inside capacitor/android-template... can we use a custom one? It would be cool

**Comment by nklayman** [comment]
@lucasfernog You probably could. Instead of running `yarn cap add android`, just copy in your custom template and possibly `yarn add -D @capacitor/android`.

**Comment by lucasfernog** [comment]
Maybe we should have one template of our own.. so we can set it up in a better way

**Comment by nklayman** [comment]
At that point we might as well make quapacitor

**Comment by eladcandroid** [comment]
@nklayman Great job!! You can use "npx cap add ios --verbose" to see what's going on..

**Comment by eladcandroid** [comment]
@nklayman Please tell me if you need any help

**Comment by nklayman** [comment]
@eladcandroid or anyone else, can you try using a capacitor/cordova plugin? Make sure to pull the latest version and create a new project.

**Comment by nothingismagick** [comment]
@eladcandroid - we'd love some help putting together the docs for this as your time permits.

File: app/lib/mode/mode-capacitor.js
Comment (lucasfernog) [inline app/lib/mode/mode-capacitor.js]: Shouldn't this be part of the template/capacitor directory?

**Review by lucasfernog — CHANGES_REQUESTED** [review]
add a `_gitignore` file to templates/capacitor with that content

**Comment by eladcandroid** [comment]
@nklayman I've followed your instructions to test the repo on mac, and got this message:
Screen Shot 2019-07-09 at 17 45 32

**Comment by nklayman** [comment]
@eladcandroid I forgot to mention that you need to switch branches before linking. Run `git checkout feat(app)--add-capacitor-support` in the `quasar/app` folder.

**Comment by eladcandroid** [comment]
@nklayman 
Checked here on Mac Mojave 10.14.5, Node version 12.6.0 and it works great including live reload! Thanks!
What's next?

**Comment by nklayman** [comment]
@eladcandroid can you try using a plugin on IOS? If that works, then we can merge this.

**Comment by eladcandroid** [comment]
> @eladcandroid can you try using a plugin on IOS? If that works, then we can merge this.

Plugins work great out of the box!
Here's a screenshot using Device.getInfo() capacitor plugin
Simulator Screen Shot - iPhone 8 - 2019-08-09 at 14 07 31

**Comment by eladcandroid** [comment]
@nklayman Any additional tasks for me? 👻

**Comment by hugetiny** [comment]
Although Capacitor is not currently compatible with some  Cordova plugins. 
Quapacitor is a perfect solution. I am looking forward to it

**Comment by eladcandroid** [comment]
@nklayman Do you need any help to finish with PR?

**Comment by nklayman** [comment]
Right now I am just waiting on review from razvan. He is extremely busy, but he should get a chance soon. If you are not deploying your app to production now you can use the fork to develop your capacitor app, and then once this gets merged you can switch off the fork and deploy your capacitor app.

**Comment by eeerrrttty** [comment]
Hello people! I do have a cordova app in production and, uol... It would be GREAT to use capacitor.

I do use some plugins:

For firebase (notifications and analytics)
Camera.

Would it be compatible?
Will the app now be native?

**Comment by nklayman** [comment]
If the plugins are compatible with capacitor they should work just fine. I know camera will work, but I'm not sure about firebase, is it a cordova or capacitor plugin? Your app will be "hybrid", meaning it is a native app, but all of your code will run in a webview.

**Comment by mesqueeb** [comment]
Lol, the name `quapacitor` is so awesome. 🤣

**Pull request merged**
